torchao/prototype/optim_8bit/adam.py

```diff
@@ -0,0 +1,200 @@
+        else:
+            out = torch.zeros_like(p)
+        return out
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+
+                grad = p.grad
+                if grad.is_sparse:
+                    raise RuntimeError("Sparse gradient is not supported")
+
+                state = self.state[p]
+
+                # State initialization
+                # state is flattened so that torch.compile won't recompile for tensors with different ndim
+                if len(state) == 0:
+                    state["step"] = torch.tensor(0.0, device=p.device)
+                    state["exp_avg"] = self._new_zero_buffer(p.view(-1), signed=True)
+                    state["exp_avg_sq"] = self._new_zero_buffer(p.view(-1), signed=False)
+                    if group["amsgrad"]:
+                        state["max_exp_avg_sq"] = self._new_zero_buffer(p.view(-1), signed=False)
+
+                state["step"] += 1
+
+                # flatten p and grad so that torch.compile won't recompile for tensors with different ndim
+                # must explicitly convert lr to Tensor since torch.compile() will treat it as a constant
+                # if it is a python float. practically, only lr is changed during training.
+                # NOTE: if lr is change at every step, moving lr to CUDA will be a bottleneck.
+                if not isinstance(group["lr"], Tensor):
+                    group["lr"] = torch.tensor(group["lr"], device=p.device)
+                single_param_adam_v1_(
+                    p.view(-1),
+                    grad.view(-1),
+                    state["step"],
+                    state["exp_avg"],
+                    state["exp_avg_sq"],
+                    state.get("max_exp_avg_sq", None),
+                    group["lr"],
+                    group["betas"][0],
+                    group["betas"][1],
+                    group["weight_decay"],
+                    group["eps"],
+                )
+
+                # with torch._fused_adam_()
+                # single_param_adam_v2_(
+                #     p.view(-1),
+                #     grad.view(-1),
+                #     state["step"],
+                #     state["exp_avg"],
+                #     state["exp_avg_sq"],
+                #     state.get("max_exp_avg_sq", None),
+                #     group["lr"],
+                #     group["betas"][0],
+                #     group["betas"][1],
+                #     group["weight_decay"],
+                #     group["eps"],
+                # )
+
+        return loss
+
+
+# this will work with any optim state tensor subclass that implements aten.lerp.Scalar and aten.copy_.default
+@torch.compile(fullgraph=True, dynamic=True)
```

Comment:
@mlazos I'm compiling the optimizer step for each param here (called vertical fusion?). This is necessary to fuse dequant+adam_step+quant together.

@msaroufim the profile is quite big, 200MB when I ran it (or I can try compressing it). Perhaps you can run and get the profile and share with @mlazos internally? The following code will profile the first 50 iterations. Lmk if you have problems running the code
```python
python train.py \
  --model "timm/vit_base_patch16_224.augreg_in21k" \
  --amp bf16 \
  --optim AdamDTQ8bit \
  --profile
```